engine-standalone-storage/src/diff.rs

```diff
@@ -0,0 +1,72 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+use std::collections::{btree_map, BTreeMap};
+
+#[derive(Debug, Default, Clone, BorshDeserialize, BorshSerialize, PartialEq, Eq)]
+/// Collection of Engine state keys which changed by executing a transaction.
+pub struct Diff(BTreeMap<Vec<u8>, DiffValue>);
+
+#[derive(Debug, Clone, BorshDeserialize, BorshSerialize, PartialEq, Eq)]
+pub enum DiffValue {
+    Modified(Vec<u8>),
+    Deleted,
+}
+
+impl DiffValue {
+    pub fn value(&self) -> Option<&[u8]> {
+        match self {
+            Self::Deleted => None,
+            Self::Modified(new_value) => Some(new_value.as_slice()),
+        }
+    }
+
+    pub fn take_value(self) -> Option<Vec<u8>> {
+        match self {
+            Self::Deleted => None,
+            Self::Modified(new_value) => Some(new_value),
+        }
+    }
+
+    pub fn try_to_bytes(&self) -> Result<Vec<u8>, std::io::Error> {
+        self.try_to_vec()
+    }
+
+    pub fn try_from_bytes(bytes: &[u8]) -> Result<Self, std::io::Error> {
+        Self::try_from_slice(bytes)
+    }
+}
+
+impl Diff {
+    /// Compose two Diffs into a single one. If there is a conflict between them
+    /// then the value from the given (`other`) Diff is kept.
+    pub fn compose(&mut self, mut other: Self) {
```

Comment:
Can we use the `append` name here? I hate to be nit picky, but if internally we are appending, I believe we should keep these verbage so that it is self describing what is going on internally.

Reply:
Done